File: temporary_not_solved/BOJ_SNS.py
###################################
import sys
file = open('./input.txt', 'r')
input = file.readline
###################################

def dfs(idx, flag, select,visited):
    global nodes,dp
    if select:
        dp[idx][1] += min(dp[idx-1][0] + 1, dp[idx-1][1] +1)
    else:
        dp[idx][0] += min(dp[idx-1])
    if flag: # if this one is early apater
        # adjs don't need to be EA
        pass
    else:
        # but in this case, adjs must be EA all
        pass


class node:
    def __init__(self, num):
        self.num = num
        self.adj = []
        # self.parent = parent
        # self.children = []
def check():
    global visit,n
    for i in range(n, 0, -1):
        if visit[i] == False:
            return False
    return True
n = int(input())
visit = [ False for _ in range(n+1) ]
nodes = {}
dp = [[0 for _ in range(2)] for _ in range(n+1)]

for i in range(n-1):
    f, t = map(int , input().split())
    if f not in nodes:
        temp1 = node(f)
        nodes[f] = temp1
    if t not in nodes[f].adj:
        nodes[f].adj.append(t)

    if t not in nodes:
        temp2 = node(t)
        nodes[t] = temp2 
    if f not in nodes[t].adj:
        nodes[t].adj.append(f)

arr = sorted(nodes.values(), key = lambda x : len(x.adj), reverse= True)
for i in arr:
    print(i.num, i.adj)
answer =0
# all_visit = 2 ** n - 1
# visit = 0
cnt = 0
for i in arr:
    answer += 1
    # if i.parent == i.num: # root
    #     cnt += 1
    # cnt += len(nodes[i.num].children)
    # if cnt >= n:
    #     print(answer)
    #     break

    # visit |= nodes[i.num].bit

    for j in nodes[i.num].adj:
        visit[j] = True
    if check():
        print(answer)
        break

###################################
correct = int(input())
if answer == correct:
    print("CORRECT")
else:
    print("WA")
###################################
# print(answer)
# for i in arr:
#     print(i.children)

# for keys in nodes:
#     print(keys, ' ' , nodes[keys].parent, ' ' , nodes[keys].children)

File: 2020/2020_07_21/BOJ_SNS.cpp
#include <bits/stdc++.h>
using namespace std;
int n;
vector<int> inp[1000001],nodes[1000001];
int dp[1000001][2];
void pruning(int cur, int prv){
    for(auto i : inp[cur]){
        if( i != prv){
           nodes[cur].push_back(i);
           pruning(i, cur);
        }
    }
}
int recur(int cur, bool visit){
    int& ret = dp[cur][visit];
    if(ret != -1) return ret;
    ret = 0;
    if(visit){
        for(auto i : nodes[cur]){
            ret += min(recur(i,0), recur(i,1));
        }
        ret++;
    }else{
        for(auto i : nodes[cur]){
            ret += recur(i,1);
        }
    }
    return ret;
}
int main(){
    scanf("%d", &n);
    for(int i=0; i<n-1; i++){
        int a,b;
        scanf("%d %d", &a, &b);
        inp[b].push_back(a);
        inp[a].push_back(b);
    }
    pruning(1,0);
    memset(dp,-1, sizeof dp);
    printf("%d", min(recur(1,0), recur(1,1)));
    return 0;
}
/* Was try with Python3 but failed */

// ###################################
// # import sys
// # file = open('./input.txt', 'r')
// # input = file.readline
// ###################################

// def pruning(cur, prv):
//     global nodes,nex
//     for nxt in nodes[cur]:
//         # print('-----------')
//         # print(f'cur : {cur}, nxt : {nxt}, prv : {prv} ')
//         # print(f'nxt_bin : {bin(nxt)} , prv_bin : {bin(prv)} , result : { nxt ^ prv } || {nxt & prv}')
//         # print('-----------')
//         if nxt ^ prv:
//             nex[cur].append(nxt)
//             pruning(nxt, cur)

// # if this one is early apater
// # adjs don't need to be EA
// # but in other case, adjs must be EA all

// def solution(cur, visit):
//     global dp,nex
//     if dp[cur][visit] != -1:
//         return dp[cur][visit]
//     dp[cur][visit] = 0
//     if not visit:
//         for i in nex[cur]:
//             dp[cur][visit] += solution(i, 1)
//     else:
//         for i in nex[cur]:
//             dp[cur][visit] += min(solution(i, 0),solution(i, 1))
//         dp[cur][visit] += 1
//     return dp[cur][visit]

// n = int(input())
// nodes = [ [] for _ in range(n+1)]
// nex = [ [] for _ in range(n+1)]
// dp = [[-1 for _ in range(2)] for _ in range(n+1)]

// for i in range(n-1):
//     f, t = map(int , input().split())
//     nodes[f].append(t)
//     nodes[t].append(f)

// pruning(1,0)

// answer = min(solution(1,0), solution(1,1))
// print(answer)

// ###################################
// # correct = int(input())
// # if answer == correct:
// #     print("CORRECT")
// # else:
// #     print("WA")
// ###################################